### PR TITLE
internal/ethapi: handle blobs in API methods

### PR DESCRIPTION
**Pull request by s1na**
EIP-4844 adds a new transaction type for blobs. Users can submit such transactions via `eth_sendRawTransaction`. In this PR we refrain from adding support to `eth_sendTransaction` and in fact it will fail if the user passes in a blob hash.

However since the chain can handle such transactions it makes sense to allow simulating them. E.g. an L2 operator should be able to simulate submitting a rollup blob and updating the L2 state. Most methods that take in a transaction object should recognize blobs.The change boils down to adding `blobVersionedHashes` and `maxFeePerBlobGas` to `TransactionArgs`. In summary:

- `eth_sendTransaction`: will fail for blob txes
- `eth_signTransaction`: will fail for blob txes

As geth moves away from user-management responsibilities, the methods that sign txes will not support the new EIP-4844 transaction types. Please use `eth_sendRawTransaction` instead. Resuming the summary:

- `eth_sendRawTransaction`: can send blob txes
- `eth_fillTransaction`: will fill in a blob tx. Note: here we simply fill in normal transaction fields + possibly `maxFeePerBlobGas` when blobs are present. One can imagine a more elaborate set-up where users can submit blobs themselves and we fill in proofs and commitments and such. Left for future PRs if desired.
- `eth_call`: can simulate blob messages
- `eth_estimateGas`: blobs have no effect here. They have a separate unit of gas which is not tunable in the transaction.

### CODE REVIEW

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -468,11 +470,15 @@ func (s *PersonalAccountAPI) SendTransaction(ctx context.Context, args Transacti
 		s.nonceLock.LockAddr(args.from())
 		defer s.nonceLock.UnlockAddr(args.from())
 	}
+	if args.BlobVersionedHashes != nil {
```
Comment (lightclient) [inline internal/ethapi/api.go:473]:
```suggestion
	if args.BlobVersionedHashes != nil || args.MaxFeePerBlobGas != nil {
```

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -468,11 +470,15 @@ func (s *PersonalAccountAPI) SendTransaction(ctx context.Context, args Transacti
...
+		return common.Hash{}, errBlobTxNotSupported
+	}
 	signed, err := s.signTransaction(ctx, &args, passwd)
 	if err != nil {
 		log.Warn("Failed transaction send attempt", "from", args.from(), "to", args.To, "value", args.Value.ToInt(), "err", err)
 		return common.Hash{}, err
 	}
+
```
Comment (lightclient) [inline internal/ethapi/api.go:481]:
```suggestion
```

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -492,6 +498,9 @@ func (s *PersonalAccountAPI) SignTransaction(ctx context.Context, args Transacti
 	if args.GasPrice == nil && (args.MaxFeePerGas == nil || args.MaxPriorityFeePerGas == nil) {
 		return nil, errors.New("missing gasPrice or maxFeePerGas/maxPriorityFeePerGas")
 	}
+	if args.BlobVersionedHashes != nil && args.MaxFeePerBlobGas == nil {
```
Comment (lightclient) [inline internal/ethapi/api.go:501]: should also have the reverse? where `args.BlobVersionedHashes == nil && args.MaxFeePerBlobGas != nil` ?

File: internal/ethapi/api.go [hunk 4]
```diff
@@ -1809,6 +1819,9 @@ func (s *TransactionAPI) SendTransaction(ctx context.Context, args TransactionAr
 		s.nonceLock.LockAddr(args.from())
 		defer s.nonceLock.UnlockAddr(args.from())
 	}
+	if args.BlobVersionedHashes != nil {
```
Comment (lightclient) [inline internal/ethapi/api.go:1822]:
```suggestion
	if args.BlobVersionedHashes != nil || args.MaxFeePerBlobGas != nil {
```
same here i think?

File: internal/ethapi/api.go [hunk 5]
```diff
@@ -1892,6 +1906,9 @@ func (s *TransactionAPI) SignTransaction(ctx context.Context, args TransactionAr
 	if args.GasPrice == nil && (args.MaxPriorityFeePerGas == nil || args.MaxFeePerGas == nil) {
 		return nil, errors.New("missing gasPrice or maxFeePerGas/maxPriorityFeePerGas")
 	}
+	if args.BlobVersionedHashes != nil && args.MaxFeePerBlobGas == nil {
```
Comment (lightclient) [inline internal/ethapi/api.go:1909]: also need reverse here?

File: internal/ethapi/api.go (hunk 1 above)
Comment (rjl493456442) [reply internal/ethapi/api.go:473]: We can define a method like 

```go
// IsEIP4844 returns an indicator if the args contains EIP4844 fields.
func (args *TransactionArgs) IsEIP4844() bool {
	return len(args.BlobVersionedHashes) != 0 || args.MaxFeePerBlobGas != nil
}
```

File: internal/ethapi/api.go (hunk 3 above)
Comment (rjl493456442) [reply internal/ethapi/api.go:501]: perhaps 

```go
if args.IsEIP4844() && (args.BlobVersionedHashes == nil || args.MaxFeePerBlobGas == nil)
```

File: internal/ethapi/api_test.go [hunk 6]
```diff
@@ -403,10 +405,30 @@ func allBlobTxs(addr common.Address, config *params.ChainConfig) []txData {
 	}
 }
 
+func newTestAccountManager(t *testing.T) (*accounts.Manager, accounts.Account) {
+	var (
+		dir        = t.TempDir()
+		am         = accounts.NewManager(&accounts.Config{InsecureUnlockAllowed: true})
+		b          = keystore.NewKeyStore(dir, 2, 1)
+		testKey, _ = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
```
Comment (KillariDev) [inline internal/ethapi/api_test.go:413]: what is this key?

File: core/state_transition.go [hunk 7]
```diff
@@ -315,8 +314,14 @@ func (st *StateTransition) preCheck() error {
 	}
 	// Check the blob version validity
 	if msg.BlobHashes != nil {
+		// The to field of a blob tx type is mandatory.
+		// However messages created through RPC (eth_call)
+		// don't have this restriction.
```
Comment (holiman) [inline core/state_transition.go:319]:
```suggestion
		// The to field of a blob tx type is mandatory, and a `BlobTx` transaction internally
		// has it as a non-nillable value, so any msg derived from blob transaction has it non-nil. 
		// However, messages created through RPC (eth_call) don't have this restriction.
```

File: internal/ethapi/testdata/eth_getBlockByHash-hash-1.json [hunk 8]
```diff
@@ -4,17 +4,17 @@
   "extraData": "0x",
   "gasLimit": "0x47e7c4",
   "gasUsed": "0x5208",
-  "hash": "0x0da274b315de8e4d5bf8717218ec43540464ef36378cb896469bb731e1d3f3cb",
+  "hash": "0xeeb5c1852740ca4bbe65b0f57baf80634ed12a2b44affe30eec3fb54437c3926",
```
Comment (holiman) [inline internal/ethapi/testdata/eth_getBlockByHash-hash-1.json:7]: What is the reason that all these json-files became different? Sorry for lazy question :)

Comment (s1na) [reply internal/ethapi/testdata/eth_getBlockByHash-hash-1.json:7]: In the test backend I baked in an account for which we have the key in the genesis config so all the hashes changed :) I use this key to test transaction signing.

File: internal/ethapi/api_test.go (hunk 6 above)
Comment (s1na) [reply internal/ethapi/api_test.go:413]: This is a key used throughout geth codebase for testing. Here I use it to sign a tx. Funds are safu :)

**Comment by s1na** [comment]
I rebased to fix the merge conflicts. Note that this PR is in accord with the spec. As you can see `GenericTransaction` in the spec takes in `maxFeePerBlobGas`, `blobVersionedHashes` (and `blobs` but I propose we remove that from the spec).

https://github.com/ethereum/execution-apis/blob/cea7eeb642052f4c2e03449dc48296def4aafc24/src/schemas/transaction.yaml#L408C15-L413

**Comment by s1na** [comment]
I changed `eth_signTransaction` so it also doesn't support blob-txes to be consistent with `eth_sendTransaction`. This is a light nudge towards using external account-management.

File: internal/ethapi/testdata/transfer_after_transfer.json [hunk 9]
```diff
@@ -0,0 +1,3 @@
+{
+
```
Comment (rjl493456442) [inline internal/ethapi/testdata/transfer_after_transfer.json:2]: Empty json file? Should we delete it?

Comment (s1na) [reply internal/ethapi/testdata/transfer_after_transfer.json]: Ah must have been something I wanted to do but changed my mind about. Removed thank you.

**Comment by holiman** [comment]
Triage discussion: 

>    eth_sendTransaction: will fail for blob txes
>    eth_signTransaction: will fail for blob txes

In a future PR, we should make these not fail: geth should be able to produce anything that is ethereum mainnet relevant/acceptable.

File: internal/ethapi/transaction_args.go [hunk 10]
```diff
@@ -53,6 +55,10 @@ type TransactionArgs struct {
 	// Introduced by AccessListTxType transaction.
 	AccessList *types.AccessList `json:"accessList,omitempty"`
 	ChainID    *hexutil.Big      `json:"chainId,omitempty"`
+
+	// Introduced by EIP-4844.
+	MaxFeePerBlobGas    *hexutil.Big  `json:"maxFeePerBlobGas"`
+	BlobVersionedHashes []common.Hash `json:"blobVersionedHashes,omitempty"`
```
Comment (karalabe) [inline internal/ethapi/transaction_args.go:61]: The json names of course need to stay the same, but IMO the internal field names should be the same as core.types.

File: internal/ethapi/transaction_args.go [hunk 11]
```diff
@@ -92,6 +98,12 @@ func (args *TransactionArgs) setDefaults(ctx context.Context, b Backend) error {
 	if args.Data != nil && args.Input != nil && !bytes.Equal(*args.Data, *args.Input) {
 		return errors.New(`both "data" and "input" are set and not equal. Please use "input" to pass transaction call data`)
 	}
+	if args.BlobVersionedHashes != nil && args.To == nil {
+		return errors.New(`blob transactions cannot have the form of a create transaction`)
+	}
+	if args.BlobVersionedHashes != nil && len(args.BlobVersionedHashes) == 0 {
+		return errors.New(`need at least 1 blob for a blob-tx`)
```
Comment (karalabe) [inline internal/ethapi/transaction_args.go:105]: blob-tx -> blob transaction

File: internal/ethapi/transaction_args.go [hunk 12]
```diff
@@ -236,9 +274,10 @@ func (args *TransactionArgs) ToMessage(globalGasCap uint64, baseFee *big.Int) (*
 		gas = globalGasCap
 	}
 	var (
-		gasPrice  *big.Int
-		gasFeeCap *big.Int
-		gasTipCap *big.Int
+		gasPrice      *big.Int
+		gasFeeCap     *big.Int
+		gasTipCap     *big.Int
+		blobGasFeeCap *big.Int
```
Comment (karalabe) [inline internal/ethapi/transaction_args.go:280]: I think blobFeeCap is the one used in types

File: internal/ethapi/api.go [hunk 13]
```diff
@@ -55,6 +55,8 @@ import (
 // allowed to produce in order to speed up calculations.
 const estimateGasErrorRatio = 0.015
 
+var errBlobTxNotSupported = errors.New("blob-tx not supported")
```
Comment (karalabe) [inline internal/ethapi/api.go:58]: no dash :D

File: internal/ethapi/transaction_args.go (hunk 10 above)
Comment (s1na) [reply internal/ethapi/transaction_args.go]: Done (this and the other comments).

File: internal/ethapi/api.go [hunk 14]
```diff
@@ -55,7 +55,7 @@ import (
 // allowed to produce in order to speed up calculations.
 const estimateGasErrorRatio = 0.015
 
-var errBlobTxNotSupported = errors.New("blob-tx not supported")
+var errBlobTxNotSupported = errors.New("signing blob transactions not supported. Use clef to sign. Submit via eth_sendRawTransaction.")
```
Comment (holiman) [inline internal/ethapi/api.go:58]: Much as I'd like to shill `clef`, I think usage-instructions do not belong in errors. If we're executing in a code-path _very_ close to the user (e.g in the main or thereabouts), we can provide usage information alongside an error-message, but putting it into an `error` is Not The Way, imo

Comment (holiman) [inline internal/ethapi/api.go:58]:
```suggestion
var errBlobTxNotSupported = errors.New("signing blob transactions not supported")
```

Comment (s1na) [reply internal/ethapi/api.go]: Roger that 🫡

**Review by holiman — APPROVED** [review]
LGTM

**Review by holiman — CHANGES_REQUESTED** [review]
Please undo the change to `tests/testdata`, I guess you committed it unintentionally.

**Comment by s1na** [comment]
> Please undo the change to tests/testdata, I guess you committed it unintentionally.

Oops, fixed.

**Review by holiman — APPROVED** [review]

**Pull request merged**